src/Caching/StackExchangeRedis/src/RedisCache.cs

```diff
@@ -49,20 +51,49 @@ public partial class RedisCache : IDistributedCache, IDisposable
     private const string AbsoluteExpirationKey = "absexp";
     private const string SlidingExpirationKey = "sldexp";
     private const string DataKey = "data";
+
+    // combined keys - same hash keys fetched constantly; avoid allocating an array each time
+    private static readonly RedisValue[] HashMembers_AbsoluteExpiration_SlidingExpiration_Data = new RedisValue[] { AbsoluteExpirationKey, SlidingExpirationKey, DataKey };
+    private static readonly RedisValue[] HashMembers_AbsoluteExpiration_SlidingExpiration = new RedisValue[] { AbsoluteExpirationKey, SlidingExpirationKey };
+
+    private static RedisValue[] GetHashFields(bool getData) => getData
+        ? HashMembers_AbsoluteExpiration_SlidingExpiration_Data
+        : HashMembers_AbsoluteExpiration_SlidingExpiration;
+
     private const long NotPresent = -1;
     private static readonly Version ServerVersionWithExtendedSetCommand = new Version(4, 0, 0);
 
-    private volatile IConnectionMultiplexer? _connection;
-    private IDatabase? _cache;
+    private volatile IDatabase? _cache;
     private bool _disposed;
     private string _setScript = SetScript;
 
     private readonly RedisCacheOptions _options;
-    private readonly string _instance;
+    private readonly RedisKey _instancePrefix;
     private readonly ILogger _logger;
 
     private readonly SemaphoreSlim _connectionLock = new SemaphoreSlim(initialCount: 1, maxCount: 1);
 
+    private long _lastConnectTicks = DateTimeOffset.UtcNow.Ticks;
+    private long _firstErrorTimeTicks;
+    private long _previousErrorTimeTicks;
+
+    //// StackExchange.Redis will also be trying to reconnect internally,
+    //// so limit how often we recreate the ConnectionMultiplexer instance
+    //// in an attempt to reconnect
+    private readonly TimeSpan ReconnectMinInterval = TimeSpan.FromSeconds(60);
+    private readonly TimeSpan ReconnectErrorThreshold = TimeSpan.FromSeconds(30);
+
+    private static DateTimeOffset ReadTimeTicks(ref long field)
+    {
+        var ticks = Interlocked.Read(ref field); // avoid torn values
```

Comment:
```suggestion
        var ticks = Volatile.Read(ref field); // avoid torn values
```
? Should be enough as the exchange below has also release semantics (due the full memory barrier).

So it's cheaper at least on machine with a stronger memory model.

Reply:
This is a fun topic; relevant discussion [here](https://github.com/dotnet/runtime/issues/67330#issuecomment-1083106423), but it is an interesting dilemma: the `Interlocked` approach *is documented* as being atomic; the `Volatile` approach is not so-documented, but currently *happens to be* atomic. It comes down to which guarantees we're happy to adopt. In this case, the actual implementation here is effectively copy-paste (with tweaks) from the Azure Cache repo - I'm inclined to leave it alone, honestly - since this is only in the "we're falling over" case, I'm not too concerned about a few extra fences.

Reply:
From you discussion:

> The `Volatile` class, on the other hand, always performs atomic operations along with the barriers, regardless of data size.

And the docs do say:

> The [Volatile](https://learn.microsoft.com/en-us/dotnet/api/system.threading.volatile?view=net-7.0) class also provides read and write operations for some 64-bit types such as [Int64](https://learn.microsoft.com/en-us/dotnet/api/system.int64?view=net-7.0) and [Double](https://learn.microsoft.com/en-us/dotnet/api/system.double?view=net-7.0). Volatile reads and writes on such 64-bit memory are atomic even on 32-bit processors, unlike regular reads and writes.

https://learn.microsoft.com/en-us/dotnet/api/system.threading.volatile?view=net-7.0#remarks

And we do rely on this elsewhere in ASP.NET.

https://github.com/dotnet/aspnetcore/blob/63c49481c2e0462c65443a9dc68ec156a46d4d55/src/Servers/Kestrel/Core/src/Internal/Infrastructure/HeartbeatManager.cs#L21

Reply:
@halter73 oh, nice; I wonder if those docs have been updated; I recall this being unclear for ages